_SPIKE_.md

```diff
@@ -0,0 +1,527 @@
+bridge) should maintain a map containing inclusion resolvers for each relation
+that is allowed to be included.
+
+Conceptually:
+
+```ts
+export class DefaultCrudRepository {
+  public inclusionResolvers: Map<string, InclusionResolver>;
+
+  // ...
+}
+```
+
+**IMPORTANT:** To support use cases where no custom repository class is created
+and applications/extensions are instantiating `DefaultCrudRepository` class
+directly, it's important to expose `inclusionResolvers` as a public property.
+
+When a repository method like `find`, `findOne` and `findById` is called to
+query the database, it should use use `inclusionResolvers` map to fetch any
+related models requested by `filter.include`.
+
+Conceptually:
+
+```ts
+export class DefaultCrudRepository<T, Relations> {
+  // ...
+
+  async find(
+    filter?: Filter<T>,
+    options?: Options,
+  ): Promise<(T & Relations)[]> {
+    const models = await ensurePromise(
+      this.modelClass.find(this.normalizeFilter(filter), options),
+    );
+    const entities = this.toEntities(models);
+    // NEW LINE ADDED TO RESOLVE INCLUDED MODELS:
+    return this.includeRelatedModels(entities, filter, options);
+  }
+
+  protected async includeRelatedModels(
+    entities: T[],
+    filter?: Filter<T>,
+    options?: Options,
+  ): Promise<(T & Relations)[]> {
+    const result = entities as (T & Relations)[];
+
+    // process relations in parallel
+    const resolveTasks = filter.include.map(async i => {
+      const relationName = i.relation;
+      const resolver = this.inclusionResolvers.get(relationName);
+      const targets = await resolver(entities, i, options);
+
+      for (const ix in result) {
+        const src = result[ix];
+        (src as AnyObject)[relationName] = targets[ix];
+      }
+    });
+
+    await Promise.all(resolveTasks);
+
+    return result;
+  }
+}
+```
+
+### 3. Model-specific repositories register inclusion resolvers
+
+Model-specific repository classes (e.g. `CategoryRepository`) should register
+inclusion resolvers for model relations, similarly to how we are creating
+relation-repository factories now.
+
+To make this process easier, relation-repository factories should provide
+`inclusionResolver` property containing the appropriate `InclusionResolver`
+implementation.
+
+Conceptually:
+
+```ts
+export class CategoryRepository extends DefaultCrudRepository {
+  products: HasManyRepositoryFactory<Product, typeof Category.prototype.id>;
+
+  constructor(
+    dataSource: juggler.DataSource,
+    productRepositoryGetter: Getter<ProductRepository>,
+  ) {
+    super(Category, dataSource);
+
+    // we already have this line to create HasManyRepository factory
+    this.products = this.createHasManyRepositoryFactoryFor(
+      'todos',
+      productRepositoryGetter,
+    );
+
+    // add this line to register inclusion resolver
+    this.registerInclusion('products', this.products.inclusionResolver);
```

Comment:
Does it make sense to have an option for `createHasManyRepositoryFactoryFor` to automate `this.registerInclusion('products', this.products.inclusionResolver)` by default?

Reply:
I did consider that idea briefly.

I find it problematic for a method called `create(...)FactoryFor` to update register inclusion resolvers, it feels like a surprising side effect to me.

I consider `createHasManyRepositoryFactoryFor` and `registerInclusion` as lower-level building blocks that should stay decoupled, and can be used to build higher-level abstractions in the future.

For example, I can imagine creating a new method called e.g. `registerHasManyRelation`, that would call both `createHasManyRepositoryFactoryFor` and `registerInclusion` under the hood and initialize any repository properties (e.g. `this.products`) along the way. 

```ts
// now
this.products = this.createHasManyRepositoryFactoryFor(
  'products',
  productRepositoryGetter,
);
this.registerInclusion('products', this.products.inclusionResolver);

// with the new API
this.registerHasManyRelation(
  'products', 
  productRepositoryGetter, 
  {allowInclusion: true} // optional
);
```

The non-trivial part is how to do this in a type-safe way - we need to let TypeScript understand that the string `'products'` must match a property `this.products` with type `HasManyRepositoryFactory`.

Anyhow, I'd like to leave this out of this initial MVP scope.

I think it may be better to spend our time on https://github.com/strongloop/loopback-next/issues/2483 (_From relation definition to REST API with no repository/controller classes_), that will allow developers to avoid manually written Repository classes altogether.

@raymondfeng If you like, I can create a follow-up task to investigate a more ergonomic API for registering relations inside Repository classes. WDYT?

Reply:
> `this.registerHasManyRelation(`

Perhaps just `this.hasMany` to keep it similar to what we have in LB3.

```ts
// inside CategoryRepository constructor
this.hasMany('products', productRepositoryGetter);
```